components/suggest/src/db.rs

```diff
@@ -0,0 +1,262 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+use std::{
+    path::Path,
+    sync::{Arc, Mutex},
+};
+
+use interrupt_support::SqlInterruptHandle;
+use rusqlite::{
+    named_params,
+    types::{FromSql, ToSql},
+    Connection, OpenFlags,
+};
+use sql_support::{open_database::open_database_with_flags, ConnExt};
+
+use crate::{
+    keyword::full_keyword, schema::SuggestConnectionInitializer, RemoteRecordId, RemoteSuggestion,
+    Result, Suggestion,
+};
+
+/// The metadata key whose value is the timestamp of the last ingested Remote
+/// Settings record.
+pub const LAST_FETCH_META_KEY: &str = "last_fetch";
+
+/// A list of IAB categories that contain non-sponsored suggestions.
+pub const NONSPONSORED_IAB_CATEGORIES: &[&str] = &["5 - Education"];
+
+/// The database connection type.
+#[derive(Clone, Copy)]
+pub(crate) enum ConnectionType {
+    ReadOnly,
+    ReadWrite,
+}
+
+impl From<ConnectionType> for OpenFlags {
+    fn from(type_: ConnectionType) -> Self {
+        match type_ {
+            ConnectionType::ReadOnly => {
+                OpenFlags::SQLITE_OPEN_URI | OpenFlags::SQLITE_OPEN_READ_ONLY
+            }
+            ConnectionType::ReadWrite => {
+                OpenFlags::SQLITE_OPEN_URI
+                    | OpenFlags::SQLITE_OPEN_CREATE
+                    | OpenFlags::SQLITE_OPEN_READ_WRITE
+            }
+        }
+    }
+}
+
+/// A thread-safe "data mapper" that wraps an SQLite connection to the Suggest
+/// database with methods for reading, writing, and deleting suggestions and
+/// metadata.
+pub(crate) struct SuggestDb {
+    conn: Mutex<Connection>,
+
+    /// An object that's used to interrupt an ongoing data mapper operation from
+    /// a different thread.
+    pub interrupt_handle: Arc<SqlInterruptHandle>,
+}
+
+impl SuggestDb {
+    /// Opens a read-only or read-write connection to a Suggest database at the
+    /// given path.
+    pub fn open(path: impl AsRef<Path>, type_: ConnectionType) -> Result<Self> {
+        let conn = open_database_with_flags(path, type_.into(), &SuggestConnectionInitializer)?;
+        Ok(Self::with_connection(conn))
+    }
+
+    fn with_connection(conn: Connection) -> Self {
+        let interrupt_handle = Arc::new(SqlInterruptHandle::new(&conn));
+        Self {
+            conn: Mutex::new(conn),
+            interrupt_handle,
+        }
+    }
+
+    /// Fetches all suggestions that match the given keyword from the database.
+    pub fn fetch_by_keyword(&self, keyword: &str) -> Result<Vec<Suggestion>> {
+        let conn = self.conn.lock().unwrap();
```

Comment:
It's a bit of a shame each of these needs to acquire the mutex, because it makes transactions tricky - eg, the store's `ingest` method can't really use a single transaction.

One idea, which adds complexity so may not be worth it would be to have a new struct, say
```
impl <'a> SuggestDbApi<'a> {
    pub fn new(conn: &'a Connection, interrupt_handle: &Arc<SqlInterruptHandle>) -> Result<Self> {
        let interrupt_scope = interrupt_handle.begin_interrupt_scope()?;
        Ok(Self { conn, interrupt_scope })
    }

    pub fn err_if_interrupted(&self) -> Result<()> {
        Ok(self.interrupt_scope.err_if_interrupted()?)
    }
...
```

with all these methods defined on it. A "reader" might then look like:

```
        let db = self.dbs()?.reader.conn.lock().unwrap();
        let reader = SuggestDbApi::new(&db, &self.dbs()?.reader.interrupt_handle)?;
```

but a writer could do:
```
        let mut db = self.dbs()?.writer.conn.lock().unwrap();
        let tx = db.transaction()?;
        let writer = SuggestDbApi::new(&tx, &self.dbs()?.writer.interrupt_handle)?;
```

and that way, `ingest()` gets to do everything in a single transaction.

Again, complexity might not be worth it, and introducing lifetimes might make people new to rust hate it, but it's just a thought I had while reading this.

Reply:
Is the mutex here used to serialize all transactions on the *same* sqlite connection? If so, I wonder if we can use a `Transaction` helper and move all those read, write, drop functions to it. Something like:

```rs
// For ingestion
let mut txn = self.db.txn(); // This will acquire a lock and begin a transaction off `self.db.conn`.
txn.clear();
txn.insert_suggestions();
txn.insert_icons();
txn.commit(); // This will commit using the underlying sqlite transaction.

// For queries
let mut txn = self.db.txn(); // Readonly ops will also require a transaction.
txn.fetch_by_keyword();
// a txn could be aborted implicitly when it goes out of the scope.
``` 

Just a thought of mine, no intent to interrupt the original conversation. :)

Reply:
> Is the mutex here used to serialize all transactions on the same sqlite connection?

The mutex is actually used to gain access to the connection - the mutex "wraps" it, as opposed to something you acquire "next to" the connection. Once you've acquired the mutex you have exclusive access to it, so transactions and commands you execute are serialized by the nature of your code.

So your suggestion is roughly what I was trying to convey - it's just that mine used many more words due to Rust's general verbosity, lifetime requirements and lack of duck-typing. I hope that if it ends up working out for Lina, she'll be able to find a way to trim some of the noise.

Reply:
@tiftran and I paired on this today, and I think we've found an elegant way to do this that combines Mark's and Nan's suggestions! I'm kind of excited about it 😁 https://github.com/mozilla/application-services/pull/5723/commits/7ddf41de4369802e7525518b97503cf26ee98216

We moved all the SQL for reading and writing into a new `SuggestDao` type, pretty much in line with Mark's sketch. This made the other call sites a little verbose at first, because we had to lock the `conn` (sometimes doing it in a nested scope, to avoid holding the lock for too long), optionally start a transaction, create a `SuggestDao`, do the work we wanted, and finally commit the transaction.

We reduced that boilerplate by adding `read` and `write` methods to `SuggestDb`, that do all the locking and transaction management for us. These methods take a closure that receives the `SuggestDao`, and return whatever the closure returns. Swift likes to use that pattern a lot, and I think it works well here, too 😊

As a nice bonus, `SuggestDao` methods that write to the database can take a `&mut self`, so we can have compile-time checking for cases like accidentally calling a write method from a `read` closure. Maybe that's going a little too far? But I like that Rust lets us model that...

How does that look?

Reply:
That looks great! I feel silly for not thinking of using a closure there :) Being able to take a `&mut self` could certainly come in handy in the future and the fact we are able to use transactions "correctly" and not use our [`unchecked_transaction`](https://github.com/mozilla/application-services/blob/aa3a34b6cb47f63e7db9ef76a4c7e2b8c25ffb68/components/support/sql/src/conn_ext.rs#L271-L278) hacks is already a big win.

Reply:
+1. Assuming the transaction will get aborted automatically when the closure returns an error, this looks pretty neat.

Reply:
Rad, thank you both! I feel like we've found new patterns for storage as we've learned what works and what doesn't—but we're definitely converging on a design 😁

I'm also happy that this lets us avoid the `unchecked_transaction` escape hatch!

> Assuming the transaction will get aborted automatically when the closure returns an error

It will!